Add new proposed API to Ancillary.Interop

jtschuster opened the pull request:
Adds the new APIs from Jeremy's proposal to Ancillary.Interop project and makes the minimum changes not break the build and tests of the VTable generator.

jtschuster committed:
Update API's and make minimum changes to build

jtschuster committed:
Update API's and make minimum changes to build

AaronRobinsonMSFT commented on src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Tests/VTableGCHandlePair.cs:
```diff
@@ -1,6 +1,7 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Runtime.InteropServices.Marshalling;
 using System.Runtime.InteropServices;
```
```suggestion
using System.Runtime.InteropServices;
using System.Runtime.InteropServices.Marshalling;
```

AaronRobinsonMSFT commented on src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/GeneratedComInterfaceAnalyzerTests.cs:
```diff
@@ -35,8 +35,9 @@ interface IFoo
-                    public partial class MyComWrappers : GeneratedComWrappersBase<ComObject>
+                    public unsafe partial class MyComWrappers : GeneratedComWrappersBase
                     {
+                        protected override ComInterfaceEntry* ComputeVtables(object obj, CreateComInterfaceFlags flags, out int count) {count = 1; return null;}
```
```suggestion
                        protected override ComInterfaceEntry* ComputeVtables(object obj, CreateComInterfaceFlags flags, out int count) {count = 0; return null;}
```

Unless we are trying to force a failure, returning `null` should mean a count of `0`.

AaronRobinsonMSFT replied:
Throughout.

AaronRobinsonMSFT commented on src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/Marshalling/VTableExtensible.cs:
```diff
@@ -0,0 +1,56 @@
+// Types that are only needed for the VTable source generator or to provide abstract concepts that the COM generator would use under the hood.
```
We need the appropriate banner on each of these new files.

```
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
```

jtschuster committed:
Update src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Tests/VTableGCHandlePair.cs

Co-authored-by: Aaron Robinson <arobins@microsoft.com>

jtschuster committed:
Update src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Tests/VTableGCHandlePair.cs

Co-authored-by: Aaron Robinson <arobins@microsoft.com>

jtschuster committed:
PR feedback

- Add license banner to new files
- Change count to 0 for null VTables

jtschuster committed:
PR feedback

- Add license banner to new files
- Change count to 0 for null VTables

jtschuster committed:
Merge branch 'vtableGenUpdate' of https://github.com/jtschuster/runtime into vtableGenUpdate

elinor-fung commented on src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Tests/ImplicitThisTests.cs:
```diff
@@ -60,7 +60,7 @@ public NativeObject(void* pointer)
                 {
                     Assert.Equal(typeof(INativeObject), type);
-                    return new VirtualMethodTableInfo((IntPtr)_pointer, new ReadOnlySpan<IntPtr>(*(void**)_pointer, 2));
+                    return new VirtualMethodTableInfo(_pointer, (void**)_pointer);
```
I think you are missing the dereference to get to the vtable.

jtschuster replied:
What's the best notation for that? I changed it to `*(void***)_pointer`, but not sure if that's the ideal cast/dereference order.

elinor-fung commented on src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/Marshalling/Base.cs:
```diff
@@ -0,0 +1,129 @@
+/// <summary>
+/// Information about a virtual method table and the unmanaged instance pointer.
+/// </summary>
+public readonly unsafe struct VirtualMethodTableInfo
```
Up to you: it might be nice for easy exposing and reading of commit history to have separate files for each class/struct like we would in the product. Might also be annoying if we have to do a bunch of re-naming as part of API approval. :)

jtschuster replied:
Good idea, I separated them out into their own files. F2 renaming in VS should make it pretty painless if we need to rename.

jtschuster committed:
PR Feedback:

- Move types to their own file and update comments
- Re-add dereference to 'this' pointer for vtable

jtschuster committed:
PR Feedback:

- Move types to their own file and update comments
- Re-add dereference to 'this' pointer for vtable

jtschuster committed:
Passing tests - not cleaned; do not review

jtschuster committed:
Merge branch 'vtableGenUpdate' of https://github.com/jtschuster/runtime into vtableGenUpdate

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into vtableGenUpdate

jtschuster committed:
Make unwrapperType optional

jtschuster committed:
Remove unused code, clean up comments, format

jtschuster commented:
ImplicitThisTests.cs generated code:

```csharp
// <auto-generated/> ManagedToNativeStubs.g.cs
namespace ComInterfaceGenerator.Tests
{
    internal unsafe partial class NativeExportsNE
    {
        internal unsafe partial class ImplicitThis
        {
            internal unsafe partial interface INativeObject
            {
                internal unsafe partial interface Native
                {
                    [System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Interop.ComInterfaceGenerator", "42.42.42.42")]
                    [System.Runtime.CompilerServices.SkipLocalsInitAttribute]
                    int INativeObject.GetData()
                    {
                        var(__this, __vtable_native) = ((System.Runtime.InteropServices.Marshalling.IUnmanagedVirtualMethodTableProvider)this).GetVirtualMethodTableInfoForKey(typeof(global::ComInterfaceGenerator.Tests.NativeExportsNE.ImplicitThis.INativeObject));
                        int __retVal;
                        {
                            __retVal = ((delegate* unmanaged<void*, int> )__vtable_native[0])(__this);
                        }

                        return __retVal;
                    }
                }
            }
        }
    }
}
namespace ComInterfaceGenerator.Tests
{
    internal unsafe partial class NativeExportsNE
    {
        internal unsafe partial class ImplicitThis
        {
            internal unsafe partial interface INativeObject
            {
                internal unsafe partial interface Native
                {
                    [System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Interop.ComInterfaceGenerator", "42.42.42.42")]
                    [System.Runtime.CompilerServices.SkipLocalsInitAttribute]
                    void INativeObject.SetData(int x)
                    {
                        var(__this, __vtable_native) = ((System.Runtime.InteropServices.Marshalling.IUnmanagedVirtualMethodTableProvider)this).GetVirtualMethodTableInfoForKey(typeof(global::ComInterfaceGenerator.Tests.NativeExportsNE.ImplicitThis.INativeObject));
                        {
                            ((delegate* unmanaged<void*, int, void> )__vtable_native[1])(__this, x);
                        }
                    }
                }
            }
        }
    }
}

// <auto-generated/> NativeInterfaces.g.cs
namespace ComInterfaceGenerator.Tests
{
    internal unsafe partial class NativeExportsNE
    {
        internal unsafe partial class ImplicitThis
        {
            internal unsafe partial interface INativeObject
            {
                [System.Runtime.InteropServices.DynamicInterfaceCastableImplementationAttribute]
                internal partial interface Native : INativeObject
                {
                }
            }
        }
    }
}

// <auto-generated/> NativeToManagedStubs.g.cs
namespace ComInterfaceGenerator.Tests
{
    internal unsafe partial class NativeExportsNE
    {
        internal unsafe partial class ImplicitThis
        {
            internal unsafe partial interface INativeObject
            {
                internal unsafe partial interface Native
                {
                    [System.Runtime.InteropServices.UnmanagedCallersOnlyAttribute]
                    internal static int ABI_GetData(void* __this_native)
                    {
                        global::ComInterfaceGenerator.Tests.NativeExportsNE.ImplicitThis.INativeObject @this;
                        int __retVal = default;
                        // Unmarshal - Convert native data to managed data.
                        @this = (global::ComInterfaceGenerator.Tests.NativeExportsNE.ImplicitThis.INativeObject)System.Runtime.InteropServices.Marshalling.UnmanagedObjectUnwrapper.GetObjectForUnmanagedWrapper<ComInterfaceGenerator.Tests.VTableGCHandlePair<ComInterfaceGenerator.Tests.NativeExportsNE.ImplicitThis.INativeObject>>(__this_native);
                        __retVal = @this.GetData();
                        return __retVal;
                    }
                }
            }
        }
    }
}
namespace ComInterfaceGenerator.Tests
{
    internal unsafe partial class NativeExportsNE
    {
        internal unsafe partial class ImplicitThis
        {
            internal unsafe partial interface INativeObject
            {
                internal unsafe partial interface Native
                {
                    [System.Runtime.InteropServices.UnmanagedCallersOnlyAttribute]
                    internal static void ABI_SetData(void* __this_native, int x)
                    {
                        global::ComInterfaceGenerator.Tests.NativeExportsNE.ImplicitThis.INativeObject @this;
                        // Unmarshal - Convert native data to managed data.
                        @this = (global::ComInterfaceGenerator.Tests.NativeExportsNE.ImplicitThis.INativeObject)System.Runtime.InteropServices.Marshalling.UnmanagedObjectUnwrapper.GetObjectForUnmanagedWrapper<ComInterfaceGenerator.Tests.VTableGCHandlePair<ComInterfaceGenerator.Tests.NativeExportsNE.ImplicitThis.INativeObject>>(__this_native);
                        @this.SetData(x);
                    }
                }
            }
        }
    }
}


// <auto-generated/> PopulateVTable.g.cs
namespace ComInterfaceGenerator.Tests
{
    internal unsafe partial class NativeExportsNE
    {
        internal unsafe partial class ImplicitThis
        {
            internal unsafe partial interface INativeObject
            {
                internal unsafe partial interface Native
                {
                    internal static void PopulateUnmanagedVirtualMethodTable(void** vtable)
                    {
                        vtable[0] = (void*)(delegate* unmanaged<void*, int> )&ABI_GetData;
                        vtable[1] = (void*)(delegate* unmanaged<void*, int, void> )&ABI_SetData;
                    }
                }
            }
        }
    }
}
```

NoImplicitThis.cs generated code:
```c#
// <auto-generated/> NativeInterfaces.g.cs
namespace ComInterfaceGenerator.Tests
{
    internal unsafe partial class NativeExportsNE
    {
        internal unsafe partial class NoImplicitThis
        {
            internal unsafe partial interface IStaticMethodTable
            {
                [System.Runtime.InteropServices.DynamicInterfaceCastableImplementationAttribute]
                internal partial interface Native : IStaticMethodTable
                {
                }
            }
        }
    }
}


// <auto-generated/> ManagedToNativeStubs.g.cs
namespace ComInterfaceGenerator.Tests
{
    internal unsafe partial class NativeExportsNE
    {
        internal unsafe partial class NoImplicitThis
        {
            internal unsafe partial interface IStaticMethodTable
            {
                internal unsafe partial interface Native
                {
                    [System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Interop.ComInterfaceGenerator", "42.42.42.42")]
                    [System.Runtime.CompilerServices.SkipLocalsInitAttribute]
                    int IStaticMethodTable.Add(int x, int y)
                    {
                        var(__this, __vtable_native) = ((System.Runtime.InteropServices.Marshalling.IUnmanagedVirtualMethodTableProvider)this).GetVirtualMethodTableInfoForKey(typeof(global::ComInterfaceGenerator.Tests.NativeExportsNE.NoImplicitThis.IStaticMethodTable));
                        int __retVal;
                        {
                            __retVal = ((delegate* unmanaged<int, int, int> )__vtable_native[0])(x, y);
                        }

                        return __retVal;
                    }
                }
            }
        }
    }
}
namespace ComInterfaceGenerator.Tests
{
    internal unsafe partial class NativeExportsNE
    {
        internal unsafe partial class NoImplicitThis
        {
            internal unsafe partial interface IStaticMethodTable
            {
                internal unsafe partial interface Native
                {
                    [System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Interop.ComInterfaceGenerator", "42.42.42.42")]
                    [System.Runtime.CompilerServices.SkipLocalsInitAttribute]
                    int IStaticMethodTable.Multiply(int x, int y)
                    {
                        var(__this, __vtable_native) = ((System.Runtime.InteropServices.Marshalling.IUnmanagedVirtualMethodTableProvider)this).GetVirtualMethodTableInfoForKey(typeof(global::ComInterfaceGenerator.Tests.NativeExportsNE.NoImplicitThis.IStaticMethodTable));
                        int __retVal;
                        {
                            __retVal = ((delegate* unmanaged<int, int, int> )__vtable_native[1])(x, y);
                        }

                        return __retVal;
                    }
                }
            }
        }
    }
}
```

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/InlinedTypes.cs:
```diff
@@ -0,0 +1,247 @@
+        /// }
+        /// </code>
+        /// </summary>
+        public static ClassDeclarationSyntax IUnknownVTableComWrappersInstance { get; } = GetIUnknownVTableComWrappersInstance();
```
As follow up, we can update the public API (already approved) to expose `ComWrappers.GetIUnknownImpl` so we don't need to inline this.

jtschuster replied:
This isn't used anywhere yet, so if we don't need it for COM generator, I'll remove it.

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Marshallers/NativeToManagedThisMarshallerFactory.cs:
```diff
@@ -30,27 +31,34 @@ private sealed class Marshaller : IMarshallingGenerator
             public ManagedTypeInfo AsNativeType(TypePositionInfo info) => new PointerTypeInfo("void*", "void*", false);
             public IEnumerable<StatementSyntax> Generate(TypePositionInfo info, StubCodeContext context)
             {
+                TypeSyntax? unwrapperType = (context as NativeToManagedStubCodeContext)?.UnwrapperType;
+                Debug.Assert(unwrapperType != null);
```
Instead of passing the unwrapper type through the context, can we pass it through the `NativeThisInfo` type (and make it not a singleton)? That way we wouldn't have a dependency on the specific underlying stub context type here.

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/TypeSymbolExtensions.cs:
```diff
@@ -300,5 +300,25 @@ public static (ImmutableArray<ITypeSymbol> TypeArguments, ImmutableArray<Nullabl
+        /// </summary>
+        public static bool IsOfType(this INamedTypeSymbol type, string typeName)
+        {
+            if (typeName.Contains('<') || typeName.Contains('+') || typeName.Contains('/'))
+                throw new ArgumentException($"Cannot handle type name in the format provided: {typeName}");
```
```suggestion
            if (typeName.Contains('<') || typeName.Contains('+') || typeName.Contains('/'))
                throw new ArgumentException("Cannot match generic or nested types", nameof(typeName));
```

jtschuster replied:
I'm not sure if that's the right message to give, because it can handle nested types and open generics as long as the delimiter is `.` instead of `+` or `/` and the generics use the `` `1`` notation instead of brackets.

jkoritzinsky replied:
Yeah, you're right. Let's do the following instead (just including the argument name in the parameter list):

```suggestion
            if (typeName.Contains('<') || typeName.Contains('+') || typeName.Contains('/'))
                throw new ArgumentException($"Cannot handle type name in the format provided: {typeName}", nameof(typeName));
```

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/TypeSymbolExtensions.cs:
```diff
@@ -300,5 +300,25 @@ public static (ImmutableArray<ITypeSymbol> TypeArguments, ImmutableArray<Nullabl
+        /// <summary>
+        /// Returns if the type is of the string representation of a type
+        /// </summary>
+        public static bool IsOfType(this INamedTypeSymbol type, string typeName)
```
Should we move all of our type-name checks to use this helper instead of comparing with `ToDisplayString()`?

jkoritzinsky replied:
We can leave this for a future PR.

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/TypeSymbolExtensions.cs:
```diff
@@ -300,5 +300,25 @@ public static (ImmutableArray<ITypeSymbol> TypeArguments, ImmutableArray<Nullabl
+        {
+            if (typeName.Contains('<') || typeName.Contains('+') || typeName.Contains('/'))
+                throw new ArgumentException($"Cannot handle type name in the format provided: {typeName}");
+            var typeNameParts = typeName.Split('.').Reverse();
```
Let's do a traditional reversed `for` loop here and avoid using LINQ.

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/PreserveSigMarshaller.cs:
```diff
@@ -0,0 +1,20 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
We can remove this file. This was only needed for a conceptual example. It's not up for API review.

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/VirtualMethodIndexAttribute.cs:
```diff
@@ -8,62 +8,63 @@
 using System.Threading.Tasks;
 
-namespace System.Runtime.InteropServices.Marshalling
+namespace System.Runtime.InteropServices.Marshalling;
```
(In this and other files) we'll probably need to switch away from file-scoped namespaces before we move this code to the `src` projects that actually ship. Should we just do that now for consistency (also reducing the diff size in this PR)?

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Tests/VTableGCHandlePair.cs:
```diff
@@ -2,15 +2,16 @@
         public static void* Allocate(TUnmanagedInterface obj)
         {
-            void** unmanaged = (void**)NativeMemory.Alloc((nuint)sizeof(void*) * (nuint)IUnmanagedVirtualMethodTableProvider.GetVirtualMethodTableLength<TUnmanagedInterface>());
+            void** unmanaged = (void**)NativeMemory.Alloc((nuint)sizeof(void*) * 2);
```
Instead of hard-coding this, pass it in as a parameter? It'll make this type a little more useful if we use it in more testing scenarios.

jtschuster replied:
Doesn't this allocation allocate for the pair, and not the VTable length? Or am I misunderstanding?

jkoritzinsky replied:
You are correct! I messed this up previously.

jtschuster committed:
PR Feedback

jtschuster committed:
Pass unwrapper type through NativeThisInfo

jtschuster committed:
Update src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/TypeSymbolExtensions.cs

Co-authored-by: Jeremy Koritzinsky <jkoritzinsky@gmail.com>

jtschuster committed:
Update src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Marshallers/NativeToManagedThisMarshallerFactory.cs

Co-authored-by: Jeremy Koritzinsky <jkoritzinsky@gmail.com>

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Marshallers/NativeToManagedThisMarshallerFactory.cs:
```diff
@@ -28,8 +28,8 @@ private sealed class Marshaller : IMarshallingGenerator
-                TypeSyntax? unwrapperType = (info.MarshallingAttributeInfo as NativeThisInfo)?.UnwrapperType;
-                Debug.Assert(unwrapperType != null);
+                Debug.Assert(info.MarshallingAttributeInfo is NativeThisInfo);
+                TypeSyntax? unwrapperType = ((NativeThisInfo)info.MarshallingAttributeInfo).UnwrapperType;
```
```suggestion
                TypeSyntax unwrapperType = ((NativeThisInfo)info.MarshallingAttributeInfo).UnwrapperType;
```

jkoritzinsky reviewed (commented):
Just one more nit.

After that's fixed and the files are moved from file-scoped namespaces back to traditional namespaces, this will be ready to merge!

jtschuster committed:
Update src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Marshallers/NativeToManagedThisMarshallerFactory.cs

Co-authored-by: Jeremy Koritzinsky <jkoritzinsky@gmail.com>

jtschuster committed:
Use traditional namespace blocks

jtschuster committed:
Add InternalsVisibleTo to Ancillary.Interop for VTableGenerator tests

jtschuster committed:
Make UnmanagedObjectUnwrapper internal

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/AssemblyInfo.cs:
```diff
@@ -0,0 +1,7 @@
+
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("ComInterfaceGenerator.Tests, PublicKey=00240000048000009400000006020000002400005253413100040000010001004b86c4cb78549b34bab61a3b1800e23bfeb5b3ec390074041536a7e3cbd97f5f04cf0f857155a8928eaa29ebfd11cfbbad3ba70efea7bda3226c6a8d370a4cd303f714486b6ebc225985a638471e6ef571cc92a4613c00b8fa65d61ccee0cbe5f36330c9a01f4183559f1bef24cc2917c6d913e3a541333a1d05d9bed22b38cb")]
+[assembly: InternalsVisibleTo("ComInterfaceGenerator.Unit.Tests, PublicKey=00240000048000009400000006020000002400005253413100040000010001004b86c4cb78549b34bab61a3b1800e23bfeb5b3ec390074041536a7e3cbd97f5f04cf0f857155a8928eaa29ebfd11cfbbad3ba70efea7bda3226c6a8d370a4cd303f714486b6ebc225985a638471e6ef571cc92a4613c00b8fa65d61ccee0cbe5f36330c9a01f4183559f1bef24cc2917c6d913e3a541333a1d05d9bed22b38cb")]
```
Why do we need these `InternalsVisibleTo` attributes?

jkoritzinsky replied:
Since they're already in Ancillary.Interop, I'd prefer we keep them public. We'll leave them in Ancillary.Interop when we move the approved APIs to the System.Runtime.InteropServices assembly.

jtschuster committed:
Make types public instead of using InternalsVisibleTo

jkoritzinsky commented:
Looks like there were some conflicts from the COM source generator skeleton PR. Can we resolve those and retrigger CI?

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into vtableGenUpdate

jtschuster committed:
Build xlf, add WrapperType to context

jtschuster committed:
Update xlf

elinor-fung reviewed (approved):
Just minor comments - nothing that I think is worth resetting CI for. Feel free to address in a future change.

jtschuster commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/InlinedTypes.cs:
```diff
@@ -0,0 +1,124 @@
+
+namespace Microsoft.Interop
+{
+    internal static class InlinedTypes
```
Yes, this is unused but may be needed later.
